cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,670 @@
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(cudf::detail::make_std_vector_async(scatter_indices, stream), "gpu.node_id");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.parent_node_ids, stream),
+            "gpu.parent_node_ids");
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "gpu.node_type");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.node_levels, stream), "gpu.node_levels");
+#endif
+  rmm::device_uvector<NodeIndexT> parent_node_ids(d_tree.parent_node_ids, stream);  // make a copy
+  auto out_pid =
+    thrust::make_zip_iterator(scatter_indices.data(), parent_node_ids.data(), node_type.data());
+  // Uses cub radix sort.
+  thrust::stable_sort_by_key(rmm::exec_policy(stream),
+                             d_tree.node_levels.data(),
+                             d_tree.node_levels.data() + num_nodes,
+                             out_pid);
+  auto gather_indices = cudf::detail::scatter_to_gather(
+    scatter_indices.begin(), scatter_indices.end(), num_nodes, stream);
+
+  rmm::device_uvector<NodeIndexT> parent_indices(num_nodes, stream);
+  // gather, except parent sentinels
+  thrust::transform(rmm::exec_policy(stream),
+                    parent_node_ids.begin(),  // first node's parent is -1
+                    parent_node_ids.end(),
+                    parent_indices.begin(),
+                    [gather_indices = gather_indices.data()] __device__(auto parent_node_id) {
+                      return (parent_node_id == -1) ? -1 : gather_indices[parent_node_id];
+                    });
+#ifdef NJP_DEBUG_PRINT
+  printf("\n");
+  print_vec(cudf::detail::make_std_vector_async(scatter_indices, stream), "gpu.node_id");
+  print_vec(cudf::detail::make_std_vector_async(parent_node_ids, stream), "gpu.parent_node_ids");
+  print_vec(cudf::detail::make_std_vector_async(node_type, stream), "gpu.node_type");
+  print_vec(cudf::detail::make_std_vector_async(d_tree.node_levels, stream), "gpu.node_levels");
+  print_vec(cudf::detail::make_std_vector_async(gather_indices, stream), "new_home");
+  print_vec(cudf::detail::make_std_vector_async(parent_indices, stream), "parent_indices");
+  print_vec(cudf::detail::make_std_vector_async(parent_node_ids, stream),
+            "parent_node_ids (restored)");
+#endif
+  // 3. Find level boundaries.
+  hostdevice_vector<size_type> level_boundaries(num_nodes + 1, stream);
+  auto level_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator<size_type>(1),
+                    thrust::make_counting_iterator<size_type>(num_nodes + 1),
+                    level_boundaries.d_begin(),
+                    [num_nodes, node_levels = d_tree.node_levels.begin()] __device__(auto index) {
+                      return index == num_nodes || node_levels[index] != node_levels[index - 1];
+                    });
+  level_boundaries.device_to_host(stream, true);
```

Comment:
Think we should only transfer the part of data that is actually relevant to us

Reply:
done.